sklearn/tests/test_base.py

```diff
@@ -314,48 +312,138 @@ def transform(self, X, y=None):
+def test_pickle_version_warning_is_issued_upon_different_version():
+    iris = datasets.load_iris()
     tree = TreeBadVersion().fit(iris.data, iris.target)
     tree_pickle_other = pickle.dumps(tree)
-    message = ("Trying to unpickle estimator TreeBadVersion from "
-               "version {0} when using version {1}. This might lead to "
-               "breaking code or invalid results. "
-               "Use at your own risk.".format("something",
-                                              sklearn.__version__))
+    message = pickle_error_message.format(estimator="TreeBadVersion",
+                                          old_version="something",
+                                          current_version=sklearn.__version__)
     assert_warns_message(UserWarning, message, pickle.loads, tree_pickle_other)
 
-    # check that not including any version also works:
+
+class TreeNoVersion(DecisionTreeClassifier):
+    def __getstate__(self):
+        return self.__dict__
+
+
+def test_pickle_version_warning_is_issued_when_no_version_info_in_pickle():
+    iris = datasets.load_iris()
     # TreeNoVersion has no getstate, like pre-0.18
     tree = TreeNoVersion().fit(iris.data, iris.target)
 
     tree_pickle_noversion = pickle.dumps(tree)
     assert_false(b"version" in tree_pickle_noversion)
-    message = message.replace("something", "pre-0.18")
-    message = message.replace("TreeBadVersion", "TreeNoVersion")
+    message = pickle_error_message.format(estimator="TreeNoVersion",
+                                          old_version="pre-0.18",
+                                          current_version=sklearn.__version__)
     # check we got the warning about using pre-0.18 pickle
     assert_warns_message(UserWarning, message, pickle.loads,
                          tree_pickle_noversion)
 
-    # check that no warning is raised for external estimators
+
+def test_pickle_version_no_warning_is_issued_with_non_sklearn_estimator():
+    iris = datasets.load_iris()
+    tree = TreeNoVersion().fit(iris.data, iris.target)
+    tree_pickle_noversion = pickle.dumps(tree)
     TreeNoVersion.__module__ = "notsklearn"
     assert_no_warnings(pickle.loads, tree_pickle_noversion)
+
+
+class DontPickleCacheMixin(object):
+    def __getstate__(self):
+        data = self.__dict__.copy()
+        data["_cache"] = None
+        return data
+
+    def __setstate__(self, state):
+        state["_restored"] = True
+        self.__dict__.update(state)
+
+
+class MultiInheritanceEstimator(BaseEstimator, DontPickleCacheMixin):
+    def __init__(self, b=5):
+        self.b = b
+        self._cache = None
+
+
+def test_pickling_when_getstate_is_overwritten_by_mixin():
+    estimator = MultiInheritanceEstimator()
+    estimator._cache = "this attribute should not be pickled"
+
+    serialized = pickle.dumps(estimator, protocol=2)
+    estimator_restored = pickle.loads(serialized)
+    assert estimator_restored.b == 5
```

Comment:
see e1417ad7a0bbe262efd3422cace9c79b85c4aefb

using py.test locally so I wasn't aware that this is an issue with nose :)